lib/http2Request.js

```diff
@@ -0,0 +1,40 @@
+/*!
+ * express
+ * Copyright(c) 2009-2013 TJ Holowaychuk
+ * Copyright(c) 2013 Roman Shtylman
+ * Copyright(c) 2014-2015 Douglas Christopher Wilson
+ * MIT Licensed
+ */
+
+'use strict';
+
+/**
+ * Module dependencies.
+ * @private
+ */
+var http2 = require('http2');
+var http2Constants = http2.constants;
+var decorator = require('./requestDecorator');
+var defineGetter = require('./utils').defineGetter;
+var http2Req = decorator(Object.create(http2.Http2ServerRequest.prototype));
+/**
+ * Override default http2 path to return only pathname
+ * as it returns full url
+ * Gets the full URL from the headers
+ * Creates a mock object for parse to cache on and read from
+ * as url in http2 refers to this.path which will results in an error
+ */
+defineGetter(http2Req, 'url', function path() {

```

Comment:
@dougwilson 
I found a better solution, in retrospect my solution was bad :)

I changed it to basically work the same, except that I rewrite the getter for url to give the raw path (avoiding the endless loop on get path / url until it explodes).

Basically what happens in http2:
get path -> overridden by the decorator to return parse object pathname
set path -> sets the full path in the headers
get url -> brings the full path from headers (instead of this.path as in the unedited Http2ServerRequest)
set url -> sets the path

Reply:
Looks better, but i think the function should be named url instead of path. We also are not going to land sugar functions that monkey patch the underlying http module regardless. Express 3 and prior used to do that and it was a disaster in the long run, because it makes assumptions based on the inner implementation of Node.js core. The HTTP2 compat layer really makes it scary to even attempt to make the implementation assumptions this makes. The docs from Node.js:

> This API targets only the public API of the HTTP/1, however many modules uses internal methods or state, and those are not supported as it is a completely different implementation.

Does Node.js think that replacing the setters is something that is sustainable? Can this be noted in the docs?

In fact https://nodejs.org/dist/latest-v8.x/docs/api/http2.html#http2_class_http2_http2serverrequest doesn't even document req.path. Why are we working around something undocumented in Node.js? Seems very dangerious.

Reply:
Oh, right, forgot the function.

You are right - this does seem a bit dangerous, perhaps some opinion from Node.js team as to what we can do here would be best. Unfortunately without this change it just doesn't work.

@benjamingr @jasnell

Reply:
@nodejs/http2